master/internal/rm/kubernetesrm/jobs.go

```diff
@@ -0,0 +1,1869 @@
+	}
+
+	pods, err := j.listPodsInAllNamespaces(context.TODO(), metaV1.ListOptions{
+		LabelSelector: fmt.Sprintf("%s=%s", determinedLabel, allocationID),
+	})
+	if err != nil {
+		return errors.Wrap(err, "error listing pods checking if they can be restored")
+	}
+
+	for _, pod := range pods.Items {
+		if _, ok := j.namespaceToPoolName[pod.Namespace]; !ok {
+			continue
+		}
+		pod := pod
+		j.podStatusCallback(&pod)
+	}
+	return nil
+}
+
+func (j *jobsService) deleteKubernetesResources(
+	jobs *batchV1.JobList, configMaps *k8sV1.ConfigMapList,
+) {
+	for _, job := range jobs.Items {
+		j.resourceRequestQueue.deleteKubernetesResources(job.Namespace, job.Name, "", "")
+	}
+
+	for _, configMap := range configMaps.Items {
+		j.resourceRequestQueue.deleteKubernetesResources(configMap.Namespace, "", configMap.Name, "")
+	}
+}
+
+func (j *jobsService) deleteDoomedKubernetesResources() error {
+	var openAllocations []model.Allocation
+	if err := db.Bun().NewSelect().Model(&openAllocations).
+		Where("end_time IS NULL").
+		Scan(context.TODO()); err != nil {
+		return errors.Wrap(err, "error querying the database for open allocations")
+	}
+	openAllocationIDs := make(set.Set[model.AllocationID])
+	for _, alloc := range openAllocations {
+		openAllocationIDs.Insert(alloc.AllocationID)
+	}
+
+	listOptions := metaV1.ListOptions{LabelSelector: determinedLabel}
+	jobs, err := j.listJobsInAllNamespaces(context.TODO(), listOptions)
+	if err != nil {
+		return errors.Wrap(err, "error listing existing pods")
+	}
+
+	toKillJobs := &batchV1.JobList{}
+	savedJobNames := make(set.Set[string])
+	for _, job := range jobs.Items {
+		if _, ok := j.namespaceToPoolName[job.Namespace]; !ok {
+			continue
+		}
+
+		resourcePool := job.Labels[resourcePoolLabel]
+		if resourcePool == "" {
+			j.syslog.Warnf("deleting job '%s' without resource pool label", job.Name)
+			toKillJobs.Items = append(toKillJobs.Items, job)
+			continue
+		}
+
+		allocationIDStr := job.Labels[determinedLabel]
+		if allocationIDStr == "" {
+			j.syslog.Warnf("deleting job '%s' without determined label (whose value is the allocation ID)", job.Name)
+			toKillJobs.Items = append(toKillJobs.Items, job)
+			continue
+		}
+		allocationID := model.AllocationID(allocationIDStr)
+
+		if !openAllocationIDs.Contains(allocationID) {
+			j.syslog.Warnf("deleting job '%s', did not find an open allocation for it", allocationID)
+			toKillJobs.Items = append(toKillJobs.Items, job)
+			continue
+		}
+
+		savedJobNames.Insert(job.Name)
+	}
+
+	configMaps, err := j.listConfigMapsInAllNamespaces(context.TODO(), listOptions)
+	if err != nil {
+		return errors.Wrap(err, "error listing existing config maps")
+	}
+	toKillConfigMaps := &k8sV1.ConfigMapList{}
+	for _, cm := range configMaps.Items {
+		if _, ok := j.namespaceToPoolName[cm.Namespace]; !ok {
+			continue
+		}
+
+		if savedJobNames.Contains(cm.Name) { // Job name is same as config map name.
+			continue
+		}
```

Comment:
there should be some better control flow alternative, but I can't think of any right now -- I know the new linter will complain at this